### ISSUE TITLE
Import existing tree registration data

### ISSUE THREAD

**Issue by sarahmonster** [a8cteam51/just-one-tree#67]
Since the CPT has now been set up in #50, we'll need to import all the existing tree data into WordPress. 

There are a few CSV import plugins that may be able to automate this process so we won't need to build anything from scratch. The trickiest bit might be getting all the existing data. 

https://github.com/ethanhinson/custom-csv-exporter
https://wordpress.org/plugins/wp-ultimate-csv-importer/
https://wordpress.org/plugins/really-simple-csv-importer/

**Comment by sarahmonster** [comment]
No luck so far:

https://github.com/ethanhinson/custom-csv-exporter: only export, no import
https://wordpress.org/plugins/wp-ultimate-csv-importer/: has some JS errors that I think are preventing the interface from working
https://wordpress.org/plugins/really-simple-csv-importer/: requires all text to be quoted; this may work but would require some manipulation of our existing data, I think

**Comment by sarahmonster** [comment]
Supra CSV: maybe, but only pro version can post to CPT
Maybe this: http://www.wpallimport.com/try/
Also tried this and it was super-confusing *and* gave me a PHP error of some kind: https://wordpress.org/plugins/csv-2-post/

**Comment by michiecat** [comment]
Here's another possibility: https://wordpress.org/plugins/import-csv-files/. It doesn't support CPTs but we can use another plugin or CLI command to handle that part after importing.

**Comment by michiecat** [comment]
Some more:

* https://wordpress.org/plugins/import-csv-files/ 
* https://wordpress.org/plugins/wp-csv-to-database/

But http://www.wpallimport.com/try/ from your list looks promising.

**Comment by sarahmonster** [comment]
Decided on wpallimport.com. Started working with fresh data today (22 February 2017). If there are any new registrations between now and launch, we'll need to import those as well, so I'm going to document the import process as much as possible.

WP All Import has templates that should allow for repeating imports with the same mappings, which should be helpful. Right now I've managed to import most of the records with minimal issues. Things I'll want to fix prior to doing a final import:

- [ ] 60 records were skipped as suspected duplicates; we may want to run these by Isabel since they actually seem like dupes (same name, address, # trees)

- [ ] there are currently two fields for newsletter opt-in: "Sign up for tree care tips campaign updates 4x yr" and "And to receive JOT/lemon news". Which one corresponds to the newsletter opt-in? Newsletters in general are a bit confusing and all over the place; there are so many different lists. Might be worth mushing them all together?

- [ ] dates in this format `16/04/2016 19:46:06` are considered invalid. `05/04/2013 20:15:37` format works as expected. I think it's getting tripped up because it's expecting a sensible date format (dd/mm/yyyy) but the records are stored as US date format (mm/dd/yyyy). To preserve date of registrations, we'll need to write a function to convert it into an understandable date format.

- [ ] we'll need a function to map zip codes to neighbourhoods (see #72)

- [ ] `contact_info_registration-type` and `contact_info_newsletter-opt-in` will need functions written so their data is stored the same way ours is (fixed by running these fields through `strtolower()` prior to import)

http://www.wpallimport.com/tour/execute-php-on-data/

**Comment by sarahmonster** [comment]
I've added some scripts & fine-tuned functionality so I think the dataset we're getting is about right now. I think the next step will be actually importing the data and having Isabel check to ensure it's right. Note that we're still waiting on the missing tree registrations, so that could take a bit more time as well.

![import-process](https://cloud.githubusercontent.com/assets/376315/23257097/325930f2-f9ba-11e6-874c-836f1aae8c7d.gif)

**Comment by sarahmonster** [comment]
Isabel provided two extra spreadsheets' worth of data today. They came in a different format so I needed to make a different template for these. 

Running all three imports on my local machine gives me a total count of 1458 registrations. Once we have confirmation from Isabel of the zip-to-neighborhood mappings in #72, I'll run the imports on the staging server, and we can send her there to check that everything looks right. 

This is from latest data at 22 February; we'll need to re-import the Google form just prior to launch, to make sure we catch any new registrations as well.

These [import template files](https://github.com/a8cteam51/just-one-tree/files/804553/templates_58b46b6ac91ef.txt) can be used to import the data on other WordPress installations, following the same set of rules I've set up.

Some notes on the data, for when we send it to Isabel (cc @wrightcj03 for reference):
- All the data from the provided spreadsheets has been preserved in case it's needed at some point; extra data (ie, data that's not being collected by our form) is saved as a custom field (so it's not visible in the UI when editing the tree posts)
- Where dates were available, I set the publish date to the available date. All the early registrations (provided in manually-produced spreadsheets) are dated at 2012-01-01, since there wasn't any date available for these records.
- If we have neighbourhoods that don't fit into our map (see #3), what do we do with those records? Should those neighbourhoods be folded into the map neighbourhoods? Is it worth bringing this up now, since I think there are neighbourhoods in my mapping CSV that don't directly map to the [neighbourhoods we'll be using in the map](https://data.sfgov.org/Geographic-Locations-and-Boundaries/Planning-Neighborhood-Groups-Map/iacs-ws63/data)?

**Comment by wrightcj03** [comment]
Mind if I email her about this now @sarahmonster ? I'll remind her to review the zip codes as well.

**Comment by sarahmonster** [comment]
Nope, you can go ahead! The import stuff is pretty smart/flexible. Would you like me to add the records & such to the staging site first so she can review them? They're just on my local machine right now.

**Comment by wrightcj03** [comment]
@sarahmonster Add the records at your leisure! I can have her review the zip codes first.

**Comment by sarahmonster** [comment]
I imported these records a while ago—should we have Isabel check the records now to be sure all the data is ok, and then we can close this issue out?

**Comment by sarahmonster** [comment]
I'm going to close this issue out—if we need to reopen, we can do that, but I think the records should be good.
